File: skills/academic-figure/templates/rf-0157-upset/python/plot.py
#!/usr/bin/env python3
"""Render exact item-set intersections as an UpSet plot."""

from __future__ import annotations

import argparse
import csv
from collections import Counter, defaultdict
from dataclasses import dataclass
from pathlib import Path
from typing import Dict, Iterable, List, Sequence, Set, Tuple

import matplotlib.pyplot as plt


SCRIPT_DIR = Path(__file__).resolve().parent
TEMPLATE_DIR = SCRIPT_DIR.parent
DEFAULT_INPUT = TEMPLATE_DIR / "demo" / "demo_membership_seed157.csv"
DEFAULT_SPEC = TEMPLATE_DIR / "demo" / "demo_set_spec.csv"
DEFAULT_OUTPUT_PREFIX = Path.cwd() / "upset_python"
MAX_SETS = 16
MAX_TOP = 50
BACKGROUND = "#FBFAF7"
INK = "#20282C"
MUTED = "#647078"
ACCENT = "#287D87"
INACTIVE = "#D8D9D5"
GRID = "#E5E2DC"


class ContractError(ValueError):
    pass


@dataclass(frozen=True)
class SetSpec:
    set_id: str
    label: str


def ordered_unique(values: Iterable[str]) -> List[str]:
    seen = set()
    result: List[str] = []
    for value in values:
        if value not in seen:
            seen.add(value)
            result.append(value)
    return result


def read_set_spec(path: Path) -> List[SetSpec]:
    if not path.is_file():
        raise ContractError(f"Set spec does not exist: {path}")
    specs: List[SetSpec] = []
    seen_ids = set()
    seen_labels = set()
    with path.open("r", encoding="utf-8-sig", newline="") as handle:
        reader = csv.DictReader(handle)
        if reader.fieldnames is None:
            raise ContractError("Set spec has no header row.")
        headers = [header.strip() for header in reader.fieldnames]
        if len(headers) != len(set(headers)):
            raise ContractError("Set spec headers must be unique.")
        missing = {"set", "label"} - set(headers)
        if missing:
            raise ContractError(f"Set spec is missing columns: {', '.join(sorted(missing))}")
        for line, row in enumerate(reader, start=2):
            if all(not str(value or "").strip() for value in row.values()):
                continue
            set_id = str(row.get("set", "") or "").strip()
            label = str(row.get("label", "") or "").strip()
            if not set_id or not label:
                raise ContractError(f"Set spec line {line}: set and label must be nonblank.")
            if set_id in seen_ids:
                raise ContractError(f"Set spec line {line}: duplicate set ID {set_id!r}.")
            if label in seen_labels:
                raise ContractError(f"Set spec line {line}: duplicate display label {label!r}.")
            seen_ids.add(set_id)
            seen_labels.add(label)
            specs.append(SetSpec(set_id, label))
    if len(specs) < 2:
        raise ContractError("UpSet requires at least two declared sets.")
    if len(specs) > MAX_SETS:
        raise ContractError(
            f"Set spec declares {len(specs)} sets; the readable limit is {MAX_SETS}. "
            "Split the set family into interpretable subsets."
        )
    return specs


def read_memberships(
    path: Path, specs: Sequence[SetSpec]
) -> Tuple[List[str], Dict[str, Set[str]], str, int]:
    if not path.is_file():
        raise ContractError(f"Membership CSV does not exist: {path}")
    known_sets = {spec.set_id for spec in specs}
    items: List[str] = []
    memberships: Dict[str, Set[str]] = defaultdict(set)
    seen_pairs = set()
    statuses: List[str] = []
    seeds: List[str] = []
    row_count = 0
    with path.open("r", encoding="utf-8-sig", newline="") as handle:
        reader = csv.DictReader(handle)
        if reader.fieldnames is None:
            raise ContractError("Membership CSV has no header row.")
        headers = [header.strip() for header in reader.fieldnames]
        if len(headers) != len(set(headers)):
            raise ContractError("Membership CSV headers must be unique.")
        missing = {"item", "set"} - set(headers)
        if missing:
            raise ContractError(
                f"Membership CSV is missing columns: {', '.join(sorted(missing))}"
            )
        for line, row in enumerate(reader, start=2):
            if all(not str(value or "").strip() for value in row.values()):
                continue
            row_count += 1
            item = str(row.get("item", "") or "").strip()
            set_id = str(row.get("set", "") or "").strip()
            if not item or not set_id:
                raise ContractError(
                    f"Line {line}: item and set must be nonblank; empty memberships are invalid."
                )
            if set_id not in known_sets:
                raise ContractError(f"Line {line}: unknown set {set_id!r}.")
            pair = (item, set_id)
            if pair in seen_pairs:
                raise ContractError(
                    f"Line {line}: duplicate membership {item!r} + {set_id!r}."
                )
            seen_pairs.add(pair)
            if item not in memberships:
                items.append(item)
            memberships[item].add(set_id)
            status = str(row.get("data_status", "") or "").strip().upper()
            seed = str(row.get("simulation_seed", "") or "").strip()
            if status:
                statuses.append(status)
            if seed:
                seeds.append(seed)

    if not items:
        raise ContractError("Membership CSV contains no memberships.")
    unused_sets = [spec.set_id for spec in specs if not any(spec.set_id in memberships[item] for item in items)]
    if unused_sets:
        raise ContractError(
            "Declared sets must not be empty; no membership found for: "
            + ", ".join(unused_sets)
        )

    unique_statuses = ordered_unique(statuses)
    unique_seeds = ordered_unique(seeds)
    if len(unique_statuses) > 1 or len(unique_seeds) > 1:
        raise ContractError("data_status and simulation_seed must each be constant.")
    if unique_statuses == ["SIMULATED"]:
        if len(statuses) != row_count or len(seeds) != row_count:
            raise ContractError(
                "Every simulated row must declare data_status=SIMULATED and one fixed seed."
            )
        try:
            seed_number = int(unique_seeds[0])
        except (IndexError, ValueError) as exc:
            raise ContractError("simulation_seed must be a positive integer.") from exc
        if seed_number <= 0 or str(seed_number) != unique_seeds[0]:
            raise ContractError("simulation_seed must be a positive integer.")
        data_note = f"SIMULATED DEMONSTRATION DATA · fixed seed {seed_number}"
    else:
        if unique_seeds:
            raise ContractError(
                "simulation_seed is only valid when data_status is SIMULATED."
            )
        data_note = "SOURCE-SUPPLIED DATA"
    return items, dict(memberships), data_note, row_count


def exact_intersections(
    items: Sequence[str], memberships: Dict[str, Set[str]], specs: Sequence[SetSpec]
) -> Counter[Tuple[int, ...]]:
    set_order = [spec.set_id for spec in specs]
    counts: Counter[Tuple[int, ...]] = Counter()
    for item in items:
        combination = tuple(
            index for index, set_id in enumerate(set_order) if set_id in memberships[item]
        )
        if not combination:
            raise ContractError(f"Item {item!r} has no declared-set membership.")
        counts[combination] += 1
    return counts


def sort_intersections(
    counts: Counter[Tuple[int, ...]], top: int
) -> Tuple[List[Tuple[Tuple[int, ...], int]], int]:
    ordered = sorted(
        counts.items(),
        key=lambda entry: (-entry[1], -len(entry[0]), entry[0]),
    )
    shown = ordered[:top]
    return shown, len(ordered) - len(shown)


def render(
    items: Sequence[str],
    memberships: Dict[str, Set[str]],
    specs: Sequence[SetSpec],
    data_note: str,
    output_prefix: Path,
    title: str,
    top: int,
    dpi: int,
) -> Tuple[Path, Path, int, int]:
    counts = exact_intersections(items, memberships, specs)
    displayed, hidden_count = sort_intersections(counts, top)
    set_sizes = [
        sum(spec.set_id in memberships[item] for item in items) for spec in specs
    ]
    intersection_count = len(displayed)
    set_count = len(specs)
    width = min(23.0, max(10.0, 4.3 + 0.58 * intersection_count))
    height = max(6.3, 3.6 + 0.38 * set_count)

    figure = plt.figure(figsize=(width, height), facecolor=BACKGROUND)
    grid = figure.add_gridspec(
        2,
        2,
        width_ratios=(2.0, max(4.5, 0.52 * intersection_count)),
        height_ratios=(2.25, max(2.3, 0.38 * set_count)),
        left=0.08,
        right=0.98,
        bottom=0.12,
        top=0.82,
        wspace=0.12,
        hspace=0.06,
    )
    info_axis = figure.add_subplot(grid[0, 0])
    bar_axis = figure.add_subplot(grid[0, 1])
    size_axis = figure.add_subplot(grid[1, 0])
    matrix_axis = figure.add_subplot(grid[1, 1], sharex=bar_axis, sharey=size_axis)

    for axis in (info_axis, bar_axis, size_axis, matrix_axis):
        axis.set_facecolor(BACKGROUND)
    info_axis.axis("off")
    info_axis.text(0.0, 0.92, "Exact-membership summary", fontsize=10.5, fontweight="bold", color=INK)
    info_axis.text(
        0.0,
        0.73,
        f"{len(items)} unique items\n{set_count} declared sets\n{len(counts)} nonzero exact intersections\n{intersection_count} displayed · {hidden_count} not displayed",
        fontsize=8.8,
        color=MUTED,
        va="top",
        linespacing=1.55,
    )

    x_positions = list(range(intersection_count))
    intersection_sizes = [count for _combination, count in displayed]
    bar_axis.bar(x_positions, intersection_sizes, width=0.72, color=ACCENT, edgecolor="white", linewidth=0.6)
    max_intersection = max(intersection_sizes)
    for x_value, count in zip(x_positions, intersection_sizes):
        bar_axis.text(x_value, count + max_intersection * 0.025, str(count), ha="center", va="bottom", fontsize=max(6.2, 8.0 - 0.06 * max(0, intersection_count - 18)), color=INK)
    bar_axis.set_ylim(0, max_intersection * 1.22)
    bar_axis.set_ylabel("Exact intersection size")
    bar_axis.set_title("Top exact intersections", loc="left", fontsize=11.5, fontweight="bold")
    bar_axis.grid(axis="y", color=GRID, linewidth=0.65)
    bar_axis.set_axisbelow(True)
    bar_axis.tick_params(axis="x", bottom=False, labelbottom=False)
    for spine in ("top", "right"):
        bar_axis.spines[spine].set_visible(False)

    y_positions = list(range(set_count))
    size_axis.barh(y_positions, set_sizes, height=0.58, color="#81979A", edgecolor="white", linewidth=0.5)
    size_axis.set_ylim(set_count - 0.5, -0.5)
    size_axis.set_yticks(y_positions)
    size_axis.set_yticklabels([spec.label for spec in specs], fontsize=max(7.0, 8.8 - 0.06 * max(0, set_count - 10)))
    size_axis.set_xlabel("Set size")
    size_axis.set_title("Set sizes", loc="left", fontsize=10.5, fontweight="bold")
    size_axis.grid(axis="x", color=GRID, linewidth=0.65)
    size_axis.set_axisbelow(True)
    max_set_size = max(set_sizes)
    size_axis.set_xlim(0, max_set_size * 1.24)
    for y_value, count in zip(y_positions, set_sizes):
        size_axis.text(count + max_set_size * 0.025, y_value, str(count), ha="left", va="center", fontsize=7.5, color=INK)
    for spine in ("top", "right"):
        size_axis.spines[spine].set_visible(False)

    matrix_axis.set_xlim(-0.5, intersection_count - 0.5)
    matrix_axis.set_ylim(set_count - 0.5, -0.5)
    for y_value in y_positions:
        if y_value % 2 == 0:
            matrix_axis.axhspan(y_value - 0.5, y_value + 0.5, color="#F2F0EA", zorder=0)
    for x_value, (combination, _count) in zip(x_positions, displayed):
        matrix_axis.scatter(
            [x_value] * set_count,
            y_positions,
            s=28,
            color=INACTIVE,
            edgecolor="none",
            zorder=1,
        )
        active_y = list(combination)
        if len(active_y) > 1:
            matrix_axis.plot(
                [x_value, x_value],
                [min(active_y), max(active_y)],
                color=INK,
                linewidth=1.5,
                zorder=2,
            )
        matrix_axis.scatter(
            [x_value] * len(active_y),
            active_y,
            s=40,
            color=INK,
            edgecolor="white",
            linewidth=0.4,
            zorder=3,
        )
    matrix_axis.set_xticks(x_positions)
    matrix_axis.set_xticklabels([str(index) for index in range(1, intersection_count + 1)], fontsize=max(5.8, 7.6 - 0.06 * max(0, intersection_count - 18)))
    matrix_axis.set_xlabel("Displayed intersection rank")
    matrix_axis.tick_params(axis="y", left=False, labelleft=False)
    for spine in ("top", "right", "left"):
        matrix_axis.spines[spine].set_visible(False)

    figure.text(0.035, 0.975, title, ha="left", va="top", fontsize=18, fontweight="bold", color=INK)
    figure.text(0.035, 0.925, data_note, ha="left", va="top", fontsize=9.0, color=MUTED)
    figure.text(
        0.035,
        0.888,
        "Each item contributes once to exactly one membership combination; bars are not inclusive overlaps.",
        ha="left",
        va="top",
        fontsize=8.2,
        color=MUTED,
    )
    figure.text(
        0.035,
        0.025,
        f"Selection: size descending, then degree descending, then declared-set index tuple ascending. "
        f"Displayed {intersection_count} of {len(counts)} nonzero exact intersections; {hidden_count} not displayed.",
        ha="left",
        va="bottom",
        fontsize=7.7,
        color=MUTED,
    )

    prefix = output_prefix.expanduser().with_suffix("")
    prefix.parent.mkdir(parents=True, exist_ok=True)
    png_path = prefix.with_suffix(".png")
    svg_path = prefix.with_suffix(".svg")
    figure.savefig(png_path, dpi=dpi, bbox_inches="tight", facecolor=BACKGROUND)
    figure.savefig(svg_path, bbox_inches="tight", facecolor=BACKGROUND)
    plt.close(figure)
    return png_path, svg_path, len(counts), hidden_count


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Render exact item-set memberships as an UpSet plot.")
    parser.add_argument("--input", type=Path, default=DEFAULT_INPUT)
    parser.add_argument("--set-spec", type=Path, default=DEFAULT_SPEC)
    parser.add_argument("--output-prefix", type=Path, default=DEFAULT_OUTPUT_PREFIX)
    parser.add_argument("--title", default="Exact set intersections")
    parser.add_argument("--top", type=int, default=12)
    parser.add_argument("--dpi", type=int, default=320)
    args = parser.parse_args()
    if not 1 <= args.top <= MAX_TOP:
        parser.error(f"--top must be between 1 and {MAX_TOP}.")
    if args.dpi < 150:
        parser.error("--dpi must be at least 150.")
    return args


def main() -> int:
    args = parse_args()
    try:
        specs = read_set_spec(args.set_spec.expanduser())
        items, memberships, data_note, row_count = read_memberships(
            args.input.expanduser(), specs
        )
        png_path, svg_path, intersection_count, hidden_count = render(
            items,
            memberships,
            specs,
            data_note,
            args.output_prefix,
            args.title,
            args.top,
            args.dpi,
        )
    except ContractError as exc:
        raise SystemExit(f"Input validation failed: {exc}") from exc
    print(
        f"Validated {row_count} unique memberships: {len(items)} items × {len(specs)} sets; "
        f"{intersection_count} nonzero exact intersections; "
        f"{intersection_count - hidden_count} displayed; {hidden_count} not displayed"
    )
    print(f"Data status: {data_note}")
    print(f"PNG: {png_path.resolve()}")
    print(f"SVG: {svg_path.resolve()}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: skills/academic-figure/templates/rf-0157-upset/r/plot.R
#!/usr/bin/env Rscript

# Exact item-set intersections as an UpSet plot using base R.

all_args <- commandArgs(trailingOnly = FALSE)
file_arg <- sub("^--file=", "", all_args[grep("^--file=", all_args)])
script_dir <- if (length(file_arg)) dirname(normalizePath(file_arg[[1]], mustWork = TRUE)) else getwd()
template_dir <- dirname(script_dir)

defaults <- list(
  input = file.path(template_dir, "demo", "demo_membership_seed157.csv"),
  set_spec = file.path(template_dir, "demo", "demo_set_spec.csv"),
  output_prefix = file.path(getwd(), "upset_r"),
  title = "Exact set intersections",
  top = 12L,
  dpi = 320L
)
max_sets <- 16L
max_top <- 50L

parse_args <- function(args) {
  result <- defaults
  aliases <- c(input = "input", `set-spec` = "set_spec", `output-prefix` = "output_prefix", title = "title", top = "top", dpi = "dpi")
  index <- 1L
  while (index <= length(args)) {
    token <- args[[index]]
    if (!startsWith(token, "--")) stop(sprintf("Unknown argument: %s", token), call. = FALSE)
    stripped <- substring(token, 3L)
    if (grepl("=", stripped, fixed = TRUE)) {
      pieces <- strsplit(stripped, "=", fixed = TRUE)[[1]]
      key <- pieces[[1]]
      value <- paste(pieces[-1], collapse = "=")
    } else {
      key <- stripped
      index <- index + 1L
      if (index > length(args)) stop(sprintf("Missing value for --%s", key), call. = FALSE)
      value <- args[[index]]
    }
    if (!key %in% names(aliases)) stop(sprintf("Unknown option: --%s", key), call. = FALSE)
    result[[aliases[[key]]]] <- value
    index <- index + 1L
  }
  result$top <- suppressWarnings(as.integer(result$top))
  result$dpi <- suppressWarnings(as.integer(result$dpi))
  if (is.na(result$top) || result$top < 1L || result$top > max_top) stop(sprintf("--top must be between 1 and %d.", max_top), call. = FALSE)
  if (is.na(result$dpi) || result$dpi < 150L) stop("--dpi must be at least 150.", call. = FALSE)
  result
}

stop_contract <- function(message) stop(sprintf("Input validation failed: %s", message), call. = FALSE)
ordered_unique <- function(values) values[!duplicated(values)]

read_character_csv <- function(path, label) {
  if (!file.exists(path) || dir.exists(path)) stop_contract(sprintf("%s does not exist: %s", label, path))
  data <- tryCatch(
    utils::read.csv(path, stringsAsFactors = FALSE, check.names = FALSE, colClasses = "character", na.strings = character()),
    error = function(error) stop_contract(conditionMessage(error))
  )
  names(data) <- trimws(names(data))
  if (!ncol(data)) stop_contract(sprintf("%s has no header row.", label))
  if (any(!nzchar(names(data))) || anyDuplicated(names(data))) stop_contract(sprintf("%s headers must be nonblank and unique.", label))
  if (nrow(data)) {
    keep <- apply(data, 1L, function(row) any(nzchar(trimws(row))))
    data <- data[keep, , drop = FALSE]
  }
  data
}

read_set_spec <- function(path) {
  data <- read_character_csv(path, "Set spec")
  missing <- setdiff(c("set", "label"), names(data))
  if (length(missing)) stop_contract(sprintf("Set spec is missing columns: %s", paste(missing, collapse = ", ")))
  data$set <- trimws(data$set)
  data$label <- trimws(data$label)
  if (any(!nzchar(data$set)) || any(!nzchar(data$label))) stop_contract("Set IDs and labels must be nonblank.")
  if (anyDuplicated(data$set) || anyDuplicated(data$label)) stop_contract("Set IDs and display labels must each be unique.")
  if (nrow(data) < 2L) stop_contract("UpSet requires at least two declared sets.")
  if (nrow(data) > max_sets) stop_contract(sprintf("Set spec declares %d sets; readable limit is %d. Split into interpretable subsets.", nrow(data), max_sets))
  data
}

read_memberships <- function(path, spec) {
  data <- read_character_csv(path, "Membership CSV")
  missing <- setdiff(c("item", "set"), names(data))
  if (length(missing)) stop_contract(sprintf("Membership CSV is missing columns: %s", paste(missing, collapse = ", ")))
  if (!nrow(data)) stop_contract("Membership CSV contains no memberships.")
  data$item <- trimws(data$item)
  data$set <- trimws(data$set)
  if (any(!nzchar(data$item)) || any(!nzchar(data$set))) stop_contract("item and set must be nonblank; empty memberships are invalid.")
  unknown <- setdiff(unique(data$set), spec$set)
  if (length(unknown)) stop_contract(sprintf("Unknown set(s): %s", paste(unknown, collapse = ", ")))
  pair_keys <- paste(data$item, data$set, sep = "\r")
  if (anyDuplicated(pair_keys)) stop_contract("Duplicate item/set memberships are not allowed.")
  unused <- setdiff(spec$set, unique(data$set))
  if (length(unused)) stop_contract(sprintf("Declared sets must not be empty: %s", paste(unused, collapse = ", ")))
  status <- if ("data_status" %in% names(data)) toupper(trimws(data$data_status)) else rep("", nrow(data))
  seed <- if ("simulation_seed" %in% names(data)) trimws(data$simulation_seed) else rep("", nrow(data))
  statuses <- ordered_unique(status[nzchar(status)])
  seeds <- ordered_unique(seed[nzchar(seed)])
  if (length(statuses) > 1L || length(seeds) > 1L) stop_contract("data_status and simulation_seed must each be constant.")
  if (identical(statuses, "SIMULATED")) {
    if (any(status != "SIMULATED") || any(!nzchar(seed))) stop_contract("Every simulated row must declare SIMULATED and one fixed seed.")
    seed_number <- suppressWarnings(as.integer(seeds[[1]]))
    if (is.na(seed_number) || seed_number <= 0L || as.character(seed_number) != seeds[[1]]) stop_contract("simulation_seed must be a positive integer.")
    data_note <- sprintf("SIMULATED DEMONSTRATION DATA · fixed seed %d", seed_number)
  } else {
    if (length(seeds)) stop_contract("simulation_seed is only valid with SIMULATED data.")
    data_note <- "SOURCE-SUPPLIED DATA"
  }
  list(data = data, items = ordered_unique(data$item), data_note = data_note)
}

build_intersections <- function(validated, spec, top) {
  combo_keys <- character(length(validated$items))
  combos <- vector("list", length(validated$items))
  for (index in seq_along(validated$items)) {
    item <- validated$items[[index]]
    member_sets <- validated$data$set[validated$data$item == item]
    member_indices <- which(spec$set %in% member_sets)
    if (!length(member_indices)) stop_contract(sprintf("Item '%s' has no declared-set membership.", item))
    combos[[index]] <- member_indices
    combo_keys[[index]] <- paste(member_indices, collapse = ",")
  }
  counts_table <- table(combo_keys)
  keys <- names(counts_table)
  counts <- as.integer(counts_table)
  combo_indices <- lapply(keys, function(key) as.integer(strsplit(key, ",", fixed = TRUE)[[1]]))
  degree <- lengths(combo_indices)
  lexical_key <- vapply(combo_indices, function(indices) paste(sprintf("%02d", indices), collapse = "-"), character(1))
  order_indices <- order(-counts, -degree, lexical_key)
  keys <- keys[order_indices]
  counts <- counts[order_indices]
  combo_indices <- combo_indices[order_indices]
  displayed_count <- min(top, length(counts))
  list(
    combos = combo_indices[seq_len(displayed_count)],
    counts = counts[seq_len(displayed_count)],
    total_intersections = length(counts),
    hidden_count = length(counts) - displayed_count,
    set_sizes = vapply(spec$set, function(set_id) length(unique(validated$data$item[validated$data$set == set_id])), integer(1))
  )
}

draw_figure <- function(intersections, validated, spec, options) {
  background <- "#FBFAF7"
  ink <- "#20282C"
  muted <- "#647078"
  accent <- "#287D87"
  inactive <- "#D8D9D5"
  grid_color <- "#E5E2DC"
  shown <- length(intersections$counts)
  set_count <- nrow(spec)
  graphics::layout(
    matrix(c(1, 2, 3, 4), nrow = 2, byrow = TRUE),
    widths = c(2.0, max(4.5, 0.52 * shown)),
    heights = c(2.25, max(2.3, 0.38 * set_count))
  )
  graphics::par(oma = c(3.1, 0.5, 4.4, 0.5), bg = background, fg = ink, family = "sans")

  graphics::par(mar = c(0.8, 0.8, 2.2, 0.8))
  graphics::plot.new()
  graphics::text(0, 0.92, "Exact-membership summary", adj = c(0, 1), cex = 1.02, font = 2)
  summary_text <- sprintf(
    "%d unique items\n%d declared sets\n%d nonzero exact intersections\n%d displayed · %d not displayed",
    length(validated$items), set_count, intersections$total_intersections,
    shown, intersections$hidden_count
  )
  graphics::text(0, 0.72, summary_text, adj = c(0, 1), cex = 0.84, col = muted)

  graphics::par(mar = c(0.8, 4.0, 2.6, 1.0))
  graphics::plot.new()
  maximum <- max(intersections$counts)
  graphics::plot.window(xlim = c(0.5, shown + 0.5), ylim = c(0, maximum * 1.23), xaxs = "i", yaxs = "i")
  y_ticks <- graphics::axTicks(2)
  graphics::abline(h = y_ticks, col = grid_color, lwd = 0.65)
  for (index in seq_len(shown)) {
    graphics::rect(index - 0.36, 0, index + 0.36, intersections$counts[[index]], col = accent, border = "white", lwd = 0.6)
    graphics::text(index, intersections$counts[[index]] + maximum * 0.03, intersections$counts[[index]], cex = max(0.58, 0.78 - 0.006 * max(0, shown - 18)))
  }
  graphics::axis(2, las = 1, cex.axis = 0.78, col = "#6D7A80")
  graphics::box(bty = "l", col = "#6D7A80")
  graphics::mtext("Exact intersection size", side = 2, line = 2.7, cex = 0.86)
  graphics::mtext("Top exact intersections", side = 3, line = 1.0, adj = 0, cex = 1.08, font = 2)

  graphics::par(mar = c(3.2, 7.0 + 0.15 * max(nchar(spec$label)), 2.1, 0.8))
  graphics::plot.new()
  maximum_set <- max(intersections$set_sizes)
  graphics::plot.window(xlim = c(0, maximum_set * 1.25), ylim = c(set_count + 0.5, 0.5), xaxs = "i", yaxs = "i")
  x_ticks <- graphics::axTicks(1)
  graphics::abline(v = x_ticks, col = grid_color, lwd = 0.65)
  for (index in seq_len(set_count)) {
    graphics::rect(0, index - 0.29, intersections$set_sizes[[index]], index + 0.29, col = "#81979A", border = "white", lwd = 0.5)
    graphics::text(intersections$set_sizes[[index]] + maximum_set * 0.025, index, intersections$set_sizes[[index]], adj = c(0, 0.5), cex = 0.72)
  }
  graphics::axis(1, cex.axis = 0.76, col = "#6D7A80")
  graphics::axis(2, at = seq_len(set_count), labels = spec$label, las = 1, tick = FALSE, cex.axis = max(0.66, 0.84 - 0.012 * max(0, set_count - 10)))
  graphics::box(bty = "l", col = "#6D7A80")
  graphics::mtext("Set size", side = 1, line = 2.2, cex = 0.84)
  graphics::mtext("Set sizes", side = 3, line = 0.9, adj = 0, cex = 1.0, font = 2)

  graphics::par(mar = c(3.2, 0.8, 2.1, 1.0))
  graphics::plot.new()
  graphics::plot.window(xlim = c(0.5, shown + 0.5), ylim = c(set_count + 0.5, 0.5), xaxs = "i", yaxs = "i")
  for (row_index in seq_len(set_count)) {
    if (row_index %% 2L == 1L) graphics::rect(0.5, row_index - 0.5, shown + 0.5, row_index + 0.5, col = "#F2F0EA", border = NA)
  }
  for (column_index in seq_len(shown)) {
    graphics::points(rep(column_index, set_count), seq_len(set_count), pch = 16, cex = 0.72, col = inactive)
    active <- intersections$combos[[column_index]]
    if (length(active) > 1L) graphics::segments(column_index, min(active), column_index, max(active), col = ink, lwd = 1.5)
    graphics::points(rep(column_index, length(active)), active, pch = 21, cex = 0.92, bg = ink, col = "white", lwd = 0.45)
  }
  graphics::axis(1, at = seq_len(shown), labels = seq_len(shown), cex.axis = max(0.56, 0.74 - 0.006 * max(0, shown - 18)), col = "#6D7A80")
  graphics::box(bty = "l", col = "#6D7A80")
  graphics::mtext("Displayed intersection rank", side = 1, line = 2.2, cex = 0.84)

  graphics::mtext(options$title, side = 3, outer = TRUE, line = 2.6, adj = 0.02, cex = 1.75, font = 2)
  graphics::mtext(validated$data_note, side = 3, outer = TRUE, line = 1.2, adj = 0.02, cex = 0.86, col = muted)
  graphics::mtext("Each item contributes once to exactly one membership combination; bars are not inclusive overlaps.", side = 3, outer = TRUE, line = 0.15, adj = 0.02, cex = 0.74, col = muted)
  graphics::mtext(sprintf("Selection: size descending, degree descending, declared-set index tuple ascending. Displayed %d of %d; %d not displayed.", shown, intersections$total_intersections, intersections$hidden_count), side = 1, outer = TRUE, line = 1.4, adj = 0.02, cex = 0.70, col = muted)
}

options <- parse_args(commandArgs(trailingOnly = TRUE))
spec <- read_set_spec(options$set_spec)
validated <- read_memberships(options$input, spec)
intersections <- build_intersections(validated, spec, options$top)
shown <- length(intersections$counts)
device_width <- min(23, max(10, 4.3 + 0.58 * shown))
device_height <- max(6.3, 3.6 + 0.38 * nrow(spec))
prefix <- file.path(dirname(options$output_prefix), tools::file_path_sans_ext(basename(options$output_prefix)))
dir.create(dirname(prefix), recursive = TRUE, showWarnings = FALSE)
png_path <- paste0(prefix, ".png")
pdf_path <- paste0(prefix, ".pdf")

grDevices::png(png_path, width = device_width, height = device_height, units = "in", res = options$dpi, bg = "#FBFAF7")
draw_figure(intersections, validated, spec, options)
grDevices::dev.off()
grDevices::pdf(pdf_path, width = device_width, height = device_height, bg = "#FBFAF7", onefile = TRUE)
draw_figure(intersections, validated, spec, options)
grDevices::dev.off()

message(sprintf("Validated %d unique memberships: %d items × %d sets; %d nonzero exact intersections; %d displayed; %d not displayed", nrow(validated$data), length(validated$items), nrow(spec), intersections$total_intersections, shown, intersections$hidden_count))
message(sprintf("Data status: %s", validated$data_note))
message(sprintf("PNG: %s", normalizePath(png_path, mustWork = TRUE)))
message(sprintf("PDF: %s", normalizePath(pdf_path, mustWork = TRUE)))
